### PR TITLE
radare2 backports

### PR DESCRIPTION
**Pull request by ret2libc**
**Your checklist for this pull request**
- [ ] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [ ] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

...

**Test plan**

...

**Closing issues**

...

### CODE REVIEW

File: librz/util/axml_resources.h [hunk 1]
```diff
@@ -0,0 +1,1593 @@
...
+	"allowNativeHeapPointerTagging",
+	"autoRevokePermissions",
+	"preserveLegacyExternalStorage",
+	"mimeGroup",
+	"gwpAsanMode",
+};
+
+size_t ANDROID_ATTRIBUTE_NAMES_SIZE = RZ_ARRAY_SIZE(ANDROID_ATTRIBUTE_NAMES);
+
+#endif // R2_AXML_RESOURCES_H
```
Comment (XVilka) [inline librz/util/axml_resources.h:1593]: A typo.

File: librz/util/axml_resources.h [hunk 2]
```diff
@@ -0,0 +1,1593 @@
+// SPDX-FileCopyrightText: 2021 keegan
+// SPDX-License-Identifier: LGPL-3.0-only
+
+#ifndef R2_AXML_RESOURCES_H
+#define R2_AXML_RESOURCES_H
+
+#include <rz_types.h>
+
+// A list of all public resources from frameworks/base/core/res/res/values/public.xml:
```
Comment (XVilka) [inline librz/util/axml_resources.h:9]: Please convert it to Doxygen.

File: librz/util/axml_resources.h [hunk 3]
```diff
@@ -0,0 +1,1593 @@
+// SPDX-FileCopyrightText: 2021 keegan
+// SPDX-License-Identifier: LGPL-3.0-only
+
+#ifndef R2_AXML_RESOURCES_H
```
Comment (XVilka) [inline librz/util/axml_resources.h:4]: A typo?

File: librz/util/axml.c [hunk 4]
```diff
@@ -0,0 +1,445 @@
...
+
+		offset += header.size - sizeof(header);
+	}
+
+	rz_buf_free(buffer);
+	free(pool);
+	free(namespace);
+	return rz_strbuf_drain(sb);
+bad:
+	eprintf("Invalid Android Binary XML\n");
```
Comment (XVilka) [inline librz/util/axml.c:438]: RZ_LOG?

File: librz/util/axml.c [hunk 5]
```diff
@@ -0,0 +1,445 @@
...
+		case TYPE_RESOURCE_MAP:
+			resource_map = (ut32 *)(data + offset);
+			resource_map_length = header.size;
+			if (resource_map_length > data_size - offset) {
+				goto bad;
+			}
+			resource_map_length /= sizeof(ut32);
+			break;
+		default:
+			eprintf("Type is not recognized: %#x\n", type);
```
Comment (XVilka) [inline librz/util/axml.c:427]: RZ_LOG()?

File: librz/util/axml.c [hunk 6]
```diff
@@ -0,0 +1,445 @@
...
+
+			pool = malloc(header_size);
+			if (!pool) {
+				goto bad;
+			}
+
+			if (rz_buf_read_at(buffer, offset, (void *)pool, header_size) != header_size) {
+				goto bad;
+			}
+		} break;
```
Comment (XVilka) [inline librz/util/axml.c:358]: Maybe put `break` inside?

File: librz/util/axml.c [hunk 7]
```diff
@@ -0,0 +1,445 @@
...
+			if (!dump_element(sb, pool, namespace, data, data_size, element,
+				    resource_map, resource_map_length, &depth, true)) {
+				free(element);
+				goto bad;
+			}
+
+			depth++;
+
+			free(element);
+		} break;
```
Comment (XVilka) [inline librz/util/axml.c:389]: `break` inside?

File: librz/util/axml.c [hunk 8]
```diff
@@ -0,0 +1,445 @@
...
+	const ut32 *resource_map = NULL;
+	ut32 resource_map_length = 0;
+	RzStrBuf *sb = NULL;
+	st32 depth = 0;
+
+	rz_return_val_if_fail(data && data_size, NULL);
+
+	RzBuffer *buffer = rz_buf_new_with_pointers(data, data_size, false);
+	if (!buffer) {
+		eprintf("Error allocating RzBuffer\n");
```
Comment (XVilka) [inline librz/util/axml.c:308]: RZ_LOG?

File: librz/util/axml.c [hunk 9]
```diff
@@ -0,0 +1,445 @@
...
+	} else {
+		rz_strbuf_appendf(sb, "</%s", name);
+	}
+
+	rz_strbuf_appendf(sb, ">\n");
+	free(name);
+	return true;
+}
+
+RZ_API char *rz_axml_decode(const ut8 *data, const ut64 data_size) {
```
Comment (XVilka) [inline librz/util/axml.c:296]: Please add Doxygen.

File: librz/util/axml.c [hunk 10]
```diff
@@ -0,0 +1,445 @@
...
+			free(key);
+			free(value);
+		}
+
+		if (count != 0) {
+			rz_strbuf_appendf(sb, " ");
+		}
+
+		for (i = 0; i < count; i++) {
+			attribute_t a = element->attributes[i];
```
Comment (XVilka) [inline librz/util/axml.c:247]: I suggest extracting what's inside the loop into a separate function.

File: librz/util/axml.c [hunk 11]
```diff
@@ -0,0 +1,445 @@
...
+	case RESOURCE_FLOAT:
+		return rz_str_newf("%f", value->data.f);
+	case RESOURCE_INT_DEC:
+		return rz_str_newf("%d", value->data.d);
+	case RESOURCE_INT_HEX:
+		return rz_str_newf("0x%x", value->data.d);
+	case RESOURCE_BOOL:
+		return rz_str_newf(value->data.d ? "true" : "false");
+	default:
+		eprintf("Resource type is not recognized: %#x\n", value->type);
```
Comment (XVilka) [inline librz/util/axml.c:215]: RZ_LOG

File: librz/util/axml.c [hunk 12]
```diff
@@ -0,0 +1,445 @@
...
+
+		if ((ut64)start16 > (ut64)data + data_size - sizeof(ut32) - n - 1) {
+			free(name);
+			return NULL;
+		}
+
+		// If UTF-16LE, decode to UTF-8 so we can print it to the screen
+		if (rz_str_utf16_to_utf8((ut8 *)name, n * 2, (const ut8 *)start16, n, true) < 0) {
+			free(name);
+			eprintf("Failed to decode UTF16-LE\n");
```
Comment (XVilka) [inline librz/util/axml.c:185]: RZ_LOG

File: librz/util/axml.c [hunk 13]
```diff
@@ -0,0 +1,445 @@
...
+RZ_PACKED(
+	typedef struct {
+		ut32 line;
+		ut32 comment;
+		ut32 prefix;
+		ut32 uri;
+	})
+namespace_t;
+
+static char *string_lookup(string_pool_t *pool, const ut8 *data, ut64 data_size, ut32 i, size_t *length) {
```
Comment (XVilka) [inline librz/util/axml.c:109]: I don't like we have here a duplicate of what's inside `librz/util/str_search.c`. But it's out of the scope and probably worth opening an issue to suggest refactoring this.

File: librz/core/cmd/cmd_print.c [hunk 14]
```diff
@@ -1275,6 +1276,16 @@ static void cmd_print_fromage(RzCore *core, const char *input, const ut8 *data,
...
 		}
 	} break;
+	case 'A': // "pFA"
+	{
+		char *s = rz_axml_decode(data, size);
+		if (s) {
+			rz_cons_printf("%s", s);
+			free(s);
+		} else {
+			eprintf("Malformed object: did you supply enough data?\ntry to change the block size (see b?)\n");
```
Comment (XVilka) [inline librz/core/cmd/cmd_print.c:1286]: RZ_LOG?

**Review by XVilka — CHANGES_REQUESTED** [review]
Please use `import-` or `dist-asan-fuzz-` branches for importing patches from r2.

File: librz/core/cmd/cmd_print.c (hunk 14 above)
Comment (ret2libc) [reply librz/core/cmd/cmd_print.c:1286]: it's used in many other places in the file. Leave it this way for now.

File: librz/util/axml.c (hunk 10 above)
Comment (ret2libc) [reply librz/util/axml.c:247]: minor thing TBH. This is a backport, not a rewrite.

**Comment by ret2libc** [comment]
https://github.com/rizinorg/rizin/pull/2620

**Pull request closed without merging**